pkg/server/endpoints/node/limiter.go

```diff
@@ -0,0 +1,185 @@
+package node
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/time/rate"
+	"google.golang.org/grpc/peer"
+)
+
+const (
+	// Max burst size
+	attestLimit int = 1
+	csrLimit    int = 500
+	jsrLimit    int = 500
+
+	AttestMsg = iota
+	CSRMsg
+	JSRMsg
+)
+
+type Limiter interface {
+	Limit(ctx context.Context, msgType, count int) error
+}
+
+// Newlimiter returns a new node api rate.Limiter
+func NewLimiter(l logrus.FieldLogger) *limiter {
+	return &limiter{
+		attestRate:   1,
+		csrRate:      500,
+		jsrRate:      500,
+		lastNotified: make(map[string]time.Time),
+		limiters:     make(map[int]map[string]*rate.Limiter),
+		log:          l,
+	}
+}
+
+type limiter struct {
+	// Allowed number of messages per second
+	attestRate rate.Limit
+	csrRate    rate.Limit
+	jsrRate    rate.Limit
+
+	lastNotified map[string]time.Time
+	limiters     map[int]map[string]*rate.Limiter
+	log          logrus.FieldLogger
+	mtx          sync.Mutex
+}
+
+// Limit enforces rate limiting policy by blocking until the specified number of messages can
+// be processed. It introspects the context in order to identify the caller. An error will be
+// returned if the context is cancelled, an invalid msgType is specified, or if the number
+// of messages exceeds the burst limit.
+func (l *limiter) Limit(ctx context.Context, msgType, count int) error {
+	callerID, err := l.callerID(ctx)
+	if err != nil {
+		return err
+	}
+
+	rl, err := l.limiterFor(msgType, callerID)
+	if err != nil {
+		return err
+	}
+
+	res := rl.ReserveN(time.Now(), count)
+	if res.Delay() > 0 {
+		l.notify(callerID, msgType)
+	}
+	if !res.OK() {
+		res.Cancel()
```

Comment:
nit: do you need to cancel if the burst limit has been exceeded?

Reply:
Not sure :) I put it there "just in case"... I've removed it.